lib/recursive-methods.rb

```diff
@@ -1,49 +1,155 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  
+  if n == 0
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  else
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse(str)
+  return str if str.length <= 1
+  reversed_str = reverse(str[1..-1]) + str[0]
```

Comment:
`str[1..-1]` creates a new array and copies elements from the original array into the new one.  So it's O(n) all by itself.